Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcContentServer.cs

```diff
@@ -153,97 +156,174 @@ public async Task<GetStatsResponse> GetStatsAsync(GetStatsRequest request, Cance
+                CopyFileResponse response = new CopyFileResponse() { Content = content, Index = chunks };
+                await responseStream.WriteAsync(response).ConfigureAwait(false);
+
+                bytes += chunkSize;
+                chunks++;
+
+                Console.WriteLine(chunks);
+            }
+
+            return (chunks, bytes);
+        }
+
+        /*
+        private async Task<(long, long)> StreamContentAsync(OpenStreamResult openStreamResult, IServerStreamWriter<CopyFileResponse> responseStream)
         {
             using (var stream = openStreamResult.Stream)
             {
-                byte[] buffer = new byte[ContentStore.Grpc.Utils.DefaultBufferSize];
+                byte[] buffer = new byte[ContentStore.Grpc.Constants.DefaultBufferSize];
                 long chunks = 0;
 
                 while (true)
                 {
                     int chunkSize = await stream.ReadAsync(buffer, 0, buffer.Length);
-                    if (chunkSize == 0) break;
+                    if (chunkSize == 0) { break; }
 
                     ByteString content = ByteString.CopyFrom(buffer, 0, chunkSize);
-                    CopyFileResponse reply = new CopyFileResponse()
+                    CopyFileResponse response = new CopyFileResponse()
                     {
-                        // TODO: Check on this code
-                        Header = new ResponseHeader(startTime, openStreamResult.Succeeded, (int)openStreamResult.Code, openStreamResult.ErrorMessage, openStreamResult.Diagnostics),
                         Content = content,
                         Index = chunks
                     };
-                    await responseStream.WriteAsync(reply);
+                    await responseStream.WriteAsync(response);
                     chunks++;
                 }
             }
         }
+        */
+
+        private async Task<(long, long)> StreamContentWithCompressionAsync(Stream reader, byte[] buffer, IServerStreamWriter<CopyFileResponse> responseStream, CancellationToken cts = default(CancellationToken))
+        {
+            Debug.Assert(!(reader is null));
+            Debug.Assert(!(responseStream is null));
+
+            long bytes = 0L;
+            long chunks = 0L;
+            using (Stream grpcStream = new BufferedWriteStream(
+                buffer,
+                async (byte[] bf, int offset, int count) =>
+                {
+                    ByteString content = ByteString.CopyFrom(bf, offset, count);
+                    CopyFileResponse response = new CopyFileResponse() { Content = content, Index = chunks };
+                    await responseStream.WriteAsync(response);
+                    bytes += count;
+                    chunks++;
+                }
+            ))
+
+            {
+                using (Stream compressionStream = new GZipStream(grpcStream, System.IO.Compression.CompressionLevel.Fastest, true))

```

Comment:
>GZipStream [](start = 54, length = 10)

It might be worth comparing the perf with the built-in gzip accorss full framework, .net Core as well as the GZip stream from ICSharpCode.SharpZipLib.GZip.